# Limpieza del DataSet: IMDB 5000 Movie

En este notebook se realiza un proceso de limpieza de datos incluyendo:

1. Carga del dataset
2. Renombrar y ordenar columnas
3. Gestión de valores nulos
4. Corrección de tipos
	1. Corrección de tipos numéricos
	2. Normalización de strings
	3. Limpieza de columnas con valores múltiples
5. Detección y eliminación de duplicados
6. Guardar dataset

## 1. Carga del DataSet

In [464]:
import pandas as pd
import numpy as np
import re

In [465]:
df = pd.read_csv("movie_metadata.csv")
df.head()

color      director_name  num_critic_for_reviews  duration  \
0  Color      James Cameron                   723.0     178.0   
1  Color     Gore Verbinski                   302.0     169.0   
2  Color         Sam Mendes                   602.0     148.0   
3  Color  Christopher Nolan                   813.0     164.0   
4    NaN        Doug Walker                     NaN       NaN   

   director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                      0.0                   855.0  Joel David Moore   
1                    563.0                  1000.0     Orlando Bloom   
2                      0.0                   161.0      Rory Kinnear   
3                  22000.0                 23000.0    Christian Bale   
4                    131.0                     NaN        Rob Walker   

   actor_1_facebook_likes        gross                           genres  ...  \
0                  1000.0  760505847.0  Action|Adventure|Fantasy|Sci-Fi  ...   
1                 40000.0  309404152.0         Action|Adventure|Fantasy  ...   
2                 11000.0  200074175.0        Action|Adventure|Thriller  ...   
3                 27000.0  448130642.0                  Action|Thriller  ...   
4                   131.0          NaN                      Documentary  ...   

  num_user_for_reviews language  country  content_rating       budget  \
0               3054.0  English      USA           PG-13  237000000.0   
1               1238.0  English      USA           PG-13  300000000.0   
2                994.0  English       UK           PG-13  245000000.0   
3               2701.0  English      USA           PG-13  250000000.0   
4                  NaN      NaN      NaN             NaN          NaN   

   title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
0      2009.0                  936.0        7.9          1.78   
1      2007.0                 5000.0        7.1          2.35   
2      2015.0                  393.0        6.8          2.35   
3      2012.0                23000.0        8.5          2.35   
4         NaN                   12.0        7.1           NaN   

  movie_facebook_likes  
0                33000  
1                    0  
2                85000  
3               164000  
4                    0  

[5 rows x 28 columns]

## 2. Renombrar y ordenar columnas

In [466]:
df.columns

Index(['color', 'director_name', 'num_critic_for_reviews', 'duration',
       'director_facebook_likes', 'actor_3_facebook_likes', 'actor_2_name',
       'actor_1_facebook_likes', 'gross', 'genres', 'actor_1_name',
       'movie_title', 'num_voted_users', 'cast_total_facebook_likes',
       'actor_3_name', 'facenumber_in_poster', 'plot_keywords',
       'movie_imdb_link', 'num_user_for_reviews', 'language', 'country',
       'content_rating', 'budget', 'title_year', 'actor_2_facebook_likes',
       'imdb_score', 'aspect_ratio', 'movie_facebook_likes'],
      dtype='object')

Las columnas están desordenadas y con nombres que a lo mejor no son aclarativos. Por eso se van a modificar las cabeceras para que sean mas leibles y descriptivas.

Esto facilitará la manipulación de datos más adelante.

In [ ]:
df.columns = [col.replace("_", " ").title() for col in df.columns]

nuevo_orden = [
    # Información básica de la película
    "Movie Title", "Title Year", "Movie Imdb Link", "Language", "Country", "Content Rating",
    # Información cinematográfica
    "Duration", "Color", "Genres", "Aspect Ratio", "Plot Keywords", "Facenumber In Poster",
    # Reparto
    "Director Name", "Actor 1 Name", "Actor 2 Name", "Actor 3 Name",
    # Dinero
    "Budget", "Gross",
    # Calidad
    "Num Voted Users", "Num User For Reviews", "Num Critic For Reviews", "Imdb Score",
    # Facebook
    "Director Facebook Likes", "Actor 1 Facebook Likes", "Actor 2 Facebook Likes",
    "Actor 3 Facebook Likes", "Cast Total Facebook Likes", "Movie Facebook Likes"
]

df = df[nuevo_orden]

## 3. Revisión de nulos y tipos

In [468]:
df.isna().sum()

Movie Title                    0
Title Year                   108
Movie Imdb Link                0
Language                      14
Country                        5
Content Rating               303
Duration                      15
Color                         19
Genres                         0
Aspect Ratio                 329
Plot Keywords                153
Facenumber In Poster          13
Director Name                104
Actor 1 Name                   7
Actor 2 Name                  13
Actor 3 Name                  23
Budget                       492
Gross                        884
Num Voted Users                0
Num User For Reviews          21
Num Critic For Reviews        50
Imdb Score                     0
Director Facebook Likes      104
Actor 1 Facebook Likes         7
Actor 2 Facebook Likes        13
Actor 3 Facebook Likes        23
Cast Total Facebook Likes      0
Movie Facebook Likes           0
dtype: int64

Viendo la cantidad de valores nulos de cada columna se pueden dividir en 3 grupos:
- Columnas sin nulos: Columnas que no contienen nulos y por lo tanto no son importantes a observar en este apartado.
- Columnas esenciales con nulos: Columnas que al tener nulos hacen que haya que eliminar la fila.
- Columnas no esenciales con nulos: Columnas en las que los nulos se les puede dar un valor por defecto.

Desglose:
- Sin nulos: Titulo, URL IMDB, Generos, Votos usuarios IMDB, Puntuacion IMDB, Facebook likes del reparto, Facebook likes pelicula.
- Esenciales: Año lanzamiento, Duracion, Nombre director, Nombre actor 1.
- No esenciales: Idiomas, Pais, Clasificacion, Color, Relacion aspecto, Palabras clave trama, Caras en poster, Nombre actor 2, Nombre actor 3, Presupuesto, Recaudacion, Numero reseñas usuarios, Numero reseñas criticos, Facebook likes director, Facebook likes actor 1, Facebook likes actor 2, Facebook likes actor 3.

In [469]:
# Número de filas actualmente en el dataset
len(df)

5043

In [470]:
# ESENCIALES:
esenciales = ["Title Year", "Duration", "Director Name", "Actor 1 Name"]
df = df.dropna(subset=esenciales)

# Número de filas tras eliminar las filas con valores nulos en columnas esenciales
len(df)

4916

Se pasa de tener 5043 filas a 4916 tras eliminar todas las filas con nulos en las columnas consideradas esenciales.

In [471]:
# NO ESENCIALES:
for col in ["Language", "Country", "Content Rating", "Color", "Plot Keywords", 
            "Actor 2 Name", "Actor 3 Name"]:
    df[col] = df[col].fillna("Unknown")
    
for col in ["Facenumber In Poster", "Num User For Reviews", "Num Critic For Reviews", 
            "Director Facebook Likes", "Actor 1 Facebook Likes", 
            "Actor 2 Facebook Likes", "Actor 3 Facebook Likes"]:
    df[col] = df[col].fillna(0)
    
for col in ["Aspect Ratio", "Budget", "Gross"]:
    df[col] = df[col].fillna(df[col].median())

df.isna().sum()

Movie Title                  0
Title Year                   0
Movie Imdb Link              0
Language                     0
Country                      0
Content Rating               0
Duration                     0
Color                        0
Genres                       0
Aspect Ratio                 0
Plot Keywords                0
Facenumber In Poster         0
Director Name                0
Actor 1 Name                 0
Actor 2 Name                 0
Actor 3 Name                 0
Budget                       0
Gross                        0
Num Voted Users              0
Num User For Reviews         0
Num Critic For Reviews       0
Imdb Score                   0
Director Facebook Likes      0
Actor 1 Facebook Likes       0
Actor 2 Facebook Likes       0
Actor 3 Facebook Likes       0
Cast Total Facebook Likes    0
Movie Facebook Likes         0
dtype: int64

Se puede comprobar que se han limpiado todos los datos tras rellenar los nulos de las columnas no esenciales con valores por defecto.

## 4. Corrección de tipos

### 4.1. Corrección de tipos numéricos

Se va a asegurar que todas las columnas numéricas sean del tipo correcto (`int` o `float`).
Esto permitirá operaciones matemáticas y estadísticas sin errores.

In [472]:
col_float = ["Aspect Ratio", "Budget", "Gross", "Imdb Score"]

for col in col_float:
    df[col] = df[col].astype(float)

In [473]:
col_int = ["Title Year", "Duration", "Facenumber In Poster", "Num Voted Users", "Num User For Reviews", "Num Critic For Reviews", "Director Facebook Likes", 
	"Actor 1 Facebook Likes", "Actor 2 Facebook Likes", "Actor 3 Facebook Likes", "Cast Total Facebook Likes", "Movie Facebook Likes"]

for col in col_int:
    df[col] = df[col].astype(int)

### 4.2. Normalización de strings

Se van a quitar los espacios al inicio y al final de los strings, y esto evita errores de coincidencia en filtros o análisis de categorías.

In [474]:
col_string = ["Movie Title", "Movie Imdb Link", "Language", "Country", "Content Rating", "Color",
              "Genres", "Plot Keywords", "Director Name", "Actor 1 Name",
              "Actor 2 Name", "Actor 3 Name", ]

for col in col_string:
    df[col] = df[col].astype(str).str.strip()

### 4.3. Limpieza de columnas con valores múltiples

Se van a limpiar los espacios y carácteres especiales (`\n`, `\t` y `"`) en "Genres" y "Plot Keywords" en cada unos de sus valores.

Antes de nada reseteamos el indice para poder recorrer correctamente el df, ya que al haber eliminado filas puede darnos error.

In [475]:
df = df = df.reset_index(drop=True)

In [476]:
# Genre
for i in range(len(df)):
    generos = df.loc[i, "Genres"]
    partes = generos.split("|")
    partes_limpias = [re.sub(r'[\n\t"]', '', g.strip()) for g in partes]
    df.loc[i, "Genres"] = "|".join(partes_limpias)

In [477]:
# Plot Keywords
for i in range(len(df)):    
    kw = df.loc[i, "Plot Keywords"]
    partes = kw.split("|")
    partes_limpias = [re.sub(r'[\n\t"]', '', k.strip()) for k in partes]
    df.loc[i, "Plot Keywords"] = "|".join(partes_limpias)

## 5. Detección y eliminación de duplicados

Se van a identificar y tratar si existen filas duplicadas.

En este caso se considera que ninguna de las columnas nombradas anteriormente actúan como id, ya que se puede repetir su valor en varias filas.

Por lo tanto, se van a buscar filas cuyos campos sean todos iguales (`fila1==fila2`)

In [478]:
duplicadas = df[df.duplicated(keep="first")]
duplicadas.sort_values(by="Movie Title")

Movie Title  Title Year  \
4826                          A Dog's Breakfast        2007   
4302           A Woman, a Gun and a Noodle Shop        2009   
2508                                   Bad Moms        2016   
2738                               Big Fat Liar        2002   
4821                                 Cat People        1982   
3380                                      Crash        2004   
4653                                 Crossroads        2002   
2717                         Dangerous Liaisons        1988   
2245                         Death at a Funeral        2007   
383                              Fantastic Four        2015   
3276                                  Footloose        1984   
2419                                   Forsaken        2015   
1419                                  From Hell        2001   
201                         Godzilla Resurgence        2016   
2563                                  Halloween        1978   
4084                               Halloween II        2009   
4182                                     Hamlet        1996   
579                                    Hercules        2014   
2440                                       Hero        2002   
4461               History of the World: Part I        1981   
2514                                Left Behind        2014   
3826                            My Soul to Take        2010   
4827                   Night of the Living Dead        1968   
3648                                          O        2001   
297                                         Pan        2015   
3052                                     Snatch        2000   
4210                           Stealing Harvard        2002   
2907                                  The Alamo        2004   
779                                The Avengers        2012   
4806                                The Calling        2014   
3408                                  The Claim        2000   
389                    The Fast and the Furious        2001   
4128                      The French Connection        1971   
4453                             The Full Monty        1997   
2480                            The Illusionist        2006   
136                        The Legend of Tarzan        2016   
4761                            The Love Letter        1998   
2712                             The Possession        2012   
185   The Twilight Saga: Breaking Dawn - Part 2        2012   
643                                Total Recall        1990   
4518                                     Trance        2013   
1194                                   Unbroken        2014   
1278                        Victor Frankenstein        2015   
2125                                Wicker Park        2004   

                                        Movie Imdb Link  Language  Country  \
4826  http://www.imdb.com/title/tt0796314/?ref_=fn_t...   English   Canada   
4302  http://www.imdb.com/title/tt1428556/?ref_=fn_t...  Mandarin    China   
2508  http://www.imdb.com/title/tt4651520/?ref_=fn_t...   English      USA   
2738  http://www.imdb.com/title/tt0265298/?ref_=fn_t...   English      USA   
4821  http://www.imdb.com/title/tt0083722/?ref_=fn_t...   English      USA   
3380  http://www.imdb.com/title/tt0375679/?ref_=fn_t...   English  Germany   
4653  http://www.imdb.com/title/tt0275022/?ref_=fn_t...   English      USA   
2717  http://www.imdb.com/title/tt0094947/?ref_=fn_t...   English      USA   
2245  http://www.imdb.com/title/tt0795368/?ref_=fn_t...   English      USA   
383   http://www.imdb.com/title/tt1502712/?ref_=fn_t...   English      USA   
3276  http://www.imdb.com/title/tt0087277/?ref_=fn_t...   English      USA   
2419  http://www.imdb.com/title/tt2271563/?ref_=fn_t...   English   Canada   
1419  http://www.imdb.com/title/tt0120681/?ref_=fn_t...   English      USA   
201   http://www.imdb.com/title/tt4262980/?ref_=fn_t...  Japanese    Japan   
2563  http://www.imdb.com/titl

44 películas duplicadas que hay que eliminar.

In [479]:
df.drop_duplicates(inplace=True)

## 6. Guardar DataSet

In [480]:
df.to_csv("imdb_5000_movies_limpio.csv", index=False)